File: packages/lightning/src/commands/bridge/toggle.ts
import type { command_opts } from '../../structures/commands.ts';
import { log_error } from '../../structures/errors.ts';
import { bridge_settings_list } from '../../structures/bridge.ts';

export async function toggle(opts: command_opts): Promise<string> {
	const bridge = await opts.lightning.data.get_bridge_by_channel(
		opts.channel,
	);

	if (!bridge) return `You are not in a bridge`;

	if (!bridge_settings_list.includes(opts.args.setting)) {
		return `That setting does not exist`;
	}

	const key = opts.args.setting as keyof typeof bridge.settings;

	bridge.settings[key] = !bridge.settings[key];

	try {
		await opts.lightning.data.edit_bridge(
			bridge,
		);
		return `Bridge settings updated successfully`;
	} catch (e) {
		log_error(e, {
			message: 'Failed to update bridge in database',
			extra: { bridge },
		});
	}
}

File: internal/commands/toggle.go
package commands

import (
	"github.com/williamhorning/lightning/internal/data"
	"github.com/williamhorning/lightning/pkg/lightning"
)

func bridgeToggle(database data.Database) *lightning.Command {
	return &lightning.Command{
		Name:        "toggle",
		Description: "toggle a setting for the bridge that this channel is part of",
		Executor: func(opts *lightning.CommandOptions) {
			if opts.Arguments["setting"] == "" {
				sendErr(missingArgumentError{argument: "setting"}, "missing argument", opts)

				return
			}

			bridge, err := database.GetBridgeByChannel(opts.ChannelID)
			if err != nil {
				sendErr(err, "failed to get bridge for channel", opts)

				return
			} else if bridge.ID == "" {
				sendErr(bridgeNotFoundError{}, "this channel is not part of a bridge", opts)

				return
			}

			if opts.Arguments["setting"] != "allow_everyone" {
				sendErr(missingArgumentError{argument: "setting"}, "invalid argument", opts)

				return
			}

			bridge.Settings.AllowEveryone = !bridge.Settings.AllowEveryone

			if err = database.CreateBridge(bridge); err != nil {
				sendErr(err, "failed to set settings", opts)

				return
			}

			if err = opts.Reply(getMessage("toggled setting!", getSettingsString(&bridge)), false); err != nil {
				sendErr(err, "failed to respond to command", opts)
			}
		},
	}
}
